server/src/main/java/io/crate/expression/scalar/HasSchemaPrivilegeFunction.java

```diff
@@ -0,0 +1,295 @@
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogTableDefinitions;
+import io.crate.types.DataTypes;
+import io.crate.user.Privilege;
+import io.crate.user.User;
+import io.crate.user.UserLookup;
+import org.elasticsearch.common.TriFunction;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+
+public class HasSchemaPrivilegeFunction extends Scalar<Boolean, Object> {
+
+    public static final String NAME = "has_schema_privilege";
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final UserLookup userLookup;
+
+    @Nullable
+    private BiFunction<UserLookup, Object, User> getUser;
+
+    private TriFunction<User, Object, String, Boolean> checkPrivilege;
+
+    private static final BiFunction<UserLookup, Object, User> userByName = (userLookup, userName) -> {
+        var user = userLookup.findUser((String) userName);
+        if (user == null) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "User %s does not exist", userName));
+        }
+        return user;
+    };
+
+    private static final BiFunction<UserLookup, Object, User> userByOid = (userLookup, userOid) -> {
+        var user = userLookup.findUser((Integer) userOid);
+        if (user == null) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "User with OID %d does not exist", userOid));
+        }
+        return user;
+    };
+
+    private static final TriFunction<User, Object, String, Boolean> checkPrivilegeBySchemaName = (user, schema, privilege) -> {
+        String schemaName = (String) schema;
+        if (PgCatalogTableDefinitions.isPgCatalogOrInformationSchema(schemaName)) {
+            return true;
+        }
```

Comment:
Actually I can move this inside `user.hasPrivilege` so that `hasPrivilege` can handle this scalar and what's done in https://github.com/crate/crate/pull/11282 and https://github.com/crate/crate/pull/11316 but as it's might lead to more refactoring decided to bring it here first. cc @jeeminso

Reply:
I think it sounds good. `isPgCatalogOrInformationSchema` seems to be used only in 4 places so wouldn't be much refactoring (unless I misunderstand?)